src/worker/helpers.go

```diff
@@ -166,46 +166,224 @@ func initOLDir(olPath string, dockerBaseImage string, newBase bool) (err error)
+	// Clean up cgroups associated with sandboxes
+	cgRoot := filepath.Join("/sys", "fs", "cgroup", filepath.Base(olPath)+"-sandboxes")
+	fmt.Printf("Attempting to clean up cgroups at %s\n", cgRoot)
+
+	cgroupErrorCount := 0
+	if files, err := os.ReadDir(cgRoot); err != nil {
+		fmt.Printf("Could not find cgroup root: %s\n", err.Error())
+	} else {
+		kill := filepath.Join(cgRoot, "cgroup.kill")
+		if err := os.WriteFile(kill, []byte(fmt.Sprintf("%d", 1)), os.ModeAppend); err != nil {
+			// Log an error if killing processes in the cgroup fails.
+			fmt.Printf("Could not kill processes in cgroup: %s\n", err.Error())
+			cgroupErrorCount += 1
+		}
+		for _, file := range files {
+			if strings.HasPrefix(file.Name(), "cg-") {
+				cg := filepath.Join(cgRoot, file.Name())
+				fmt.Printf("Attempting to remove %s\n", cg)
+				if err := syscall.Rmdir(cg); err != nil {
+					// Print an error if removing a cgroup fails.
+					fmt.Printf("could not remove cgroup: %s", err.Error())
+					cgroupErrorCount += 1
+				}
+			}
+		}
+		if err := syscall.Rmdir(cgRoot); err != nil {
+			// Log an error if removing the cgroup root directory fails.
+			fmt.Printf("could not remove cgroup root: %s", err.Error())
+			cgroupErrorCount += 1
+		}
+	}
+	sandboxErrorCount := 0
+	// Clean up mounts associated with sandboxes
+	dirName := filepath.Join(olPath, "worker", "root-sandboxes")
+	fmt.Printf("Attempting to clean up mounts at %s\n", dirName)
+
+	if files, err := os.ReadDir(dirName); err != nil {
+		fmt.Printf("Could not find mount root: %s", err.Error())
+	} else {
+		for _, file := range files {
+			path := filepath.Join(dirName, file.Name())
+			fmt.Printf("Attempting to unmount %s\n", path)
+			if err := syscall.Unmount(path, syscall.MNT_DETACH); err != nil {
+				// Return an error if unmounting fails.
+				sandboxErrorCount += 1
+				fmt.Printf("Could not unmount: %s\n", err.Error())
+			}
+			if err := syscall.Rmdir(path); err != nil {
+				// Return an error if removing the mount directory fails.
+				sandboxErrorCount += 1
+				fmt.Printf("Could not remove mount dir: %s\n", err.Error())
+			}
+		}
+	}
+
+	// If we encountered any error while cleaning up the CGroup or the sandboxes
+	// return an error
+	if cgroupErrorCount != 0 || sandboxErrorCount != 0 {
+		if cgroupErrorCount != 0 {
+			fmt.Printf("%d error(s) while cleaning up cgroup.\n", cgroupErrorCount)
+		}
+		if sandboxErrorCount != 0 {
+			fmt.Printf("%d error(s) while cleaning up sandboxes.\n", sandboxErrorCount)
+		}
+		fmt.Printf("You can try to rerun the cleanup process again later.\n")
+		return fmt.Errorf("%d error(s) while cleaning up cgroup and %d error(s) while cleaning up sandbox", cgroupErrorCount, sandboxErrorCount)
+	}
+
+	// Attempt to unmount the main mount directory
+	if err := syscall.Unmount(dirName, syscall.MNT_DETACH); err != nil {
+		// Log an error if unmounting the main directory fails.
+		fmt.Printf("Could not unmount %s: %s\n", dirName, err.Error())
```

Comment:
Because when system shuts down while OL is running all those mounts will be unmounted, and calling that will result in an error. But I do agree this might not be an optimal implementation as we might encounter other errors while unmounting.